### ISSUE TITLE
 3-stage Standardization Process Proposal in staging proposal

### ISSUE THREAD

**Issue by evanp** [swicg/potential-charters#33]
This should be clearly one option. I don't think we should adopt this wholesale.

**Comment by gobengo** [comment]
Sorry, I don't understand the issue.

**Comment by dmitrizagidulin** [comment]
I'm not sure I understand what @evanp means either.

**Comment by evanp** [comment]
@aschrijver architected an [entire structure for developing standards](https://socialhub.activitypub.rocks/t/3-stage-standards-process-guaranteeing-an-open-and-decentralized-ecosystem/3602).

I don't think we should adopt it wholesale. Mentioning it in the staging document, we should make it clear that it's one path to Stage 0, but not the only one.

**Comment by aschrijver** [comment]
Indeed, thank you. The 3-stage process sketches an overall approach, an idea, for an open, bottom-up and grassroots ecosystem that facilitates the gradual adoption of contributions based on their merits.

Here's the conceptual diagram copied [from the thread](https://socialhub.activitypub.rocks/t/3-stage-standards-process-guaranteeing-an-open-and-decentralized-ecosystem/3602/40?u=aschrijver):

[![fep-process-role-in-fediverse_excalidraw](https://github.com/user-attachments/assets/1b0fb16e-935e-44f4-816d-8448f9ab9289)](https://socialhub.activitypub.rocks/t/3-stage-standards-process-guaranteeing-an-open-and-decentralized-ecosystem/3602/40?u=aschrijver)

<small>1st stage is the decentralized ecosystem, where SocialHub is part of and only facilitates the FEP process, which is the 2nd stage. The FEP informs the W3C who cherry-pick both from the FEP and from the ecosystem. The ecosystem either participates in the FEP or the W3C or both. Only the W3C gives formal guidance. FEP and ecosystem have mixed informative/formal practices.</small>

The scope of our decentralized technology base (AS/AP et al) is so vast that in Stage 1 numerous parties feed into the process, as a funnel. Thus specically supporting a decentralized ecosystem of otherwise independent communities, dev portals and (domain-specific) initiatives.

**Comment by bumblefudge** [comment]
For the record, Arnold's document has been listed prominently as "Prior Art" [since my initial commit](https://github.com/swicg/potential-charters/commit/e05cc44b6ab89be3ae83f87b7802f9cc33b2c3e8#diff-d81012efbd769bfcf839266111e957683692c0665a78aeb25cf5b5370440e6c6R97) and was very much the inspiration and starting point for this document!  Arnold's stages 1 and 2 are explicitly recommended as the default way of doing stage 0 here, and stages 1-4 are just thinner slices of Arnold's stage 3 (that people not interested in Arnold's Stage 3 can safely ignore, as it changes nothing about 1 and 2).

Anyone who wants to PR in language re-arranging or wordsmithing the recommendation is welcome to do so!  I'm not sure what would "make it clear that it's one path to Stage 0, but not the only one" better than status quo, as that was my original intention in the current wording.

As for "adopting wholesale", I have been saying all along we should ratify a CG charter first, then fine-tune this until it has consensus (or replace it with another candidate solution!), and then move on to WG candidate solutions.  So, PRs and/or equivalent alternate documents welcome?

**Comment by evanp** [comment]
I think this should be "staging process" rather than "CG Charter".

**Comment by TallTed** [comment]
[The diagram](https://github.com/swicg/potential-charters/issues/33#issuecomment-2463777680) suggests that the "W3C SocialCG" produces "W3C Recommendations" and "W3C Notes". This is not accurate. W3C Working Groups (WGs) produce W3C Recommendations and W3C Notes. W3C Community Groups (CGs) produce CG Reports, and sometimes Proposed Charters for WGs.

**Comment by aschrijver** [comment]
Yes, you are right @TallTed .. I just drew SocialCG in the diagram, as it was unknown at the time what direction things would go. If anyone wants to adjust the diagram.. I exported it as PNG with the diagram definition embedded, so it can be easily opened in Excalidraw for editing.

**Comment by TallTed** [comment]
Going forward, it would be good to change from PNG to SVG, for a variety of reasons, not least being the ability to embed a description for use by tools of readers with visual impairments.

**Comment by aschrijver** [comment]
That would be the best option. But when using Excalidraw test it for your purposes. At the time the SVG was only useful for hosting the diagram on your site, so that you could provide the sketch font and dark/light mode (e.g. with CSS).

PS. In my comment above the alt-text is present. I copied it between `<small>` as would work on Discourse, yet GH strips that out. The alt-text is:

> 1st stage is the decentralized ecosystem, where SocialHub is part of and only facilitates the FEP process, which is the 2nd stage. The FEP informs the W3C who cherry-pick both from the FEP and from the ecosystem. The ecosystem either participates in the FEP or the W3C or both. Only the W3C gives formal guidance. FEP and ecosystem have mixed informative/formal practices.

**Comment by aschrijver** [comment]
While this issue is still open, a more elaborate [staged process](https://swicg.github.io/charters/stage-process) is being proposed. I want to mention feedback wrt this document, by @silverpill and also by me, on the fediverse..

@silverpill [writes](https://web.archive.org/web/20260830162413/https://mitra.social/post/01a03a37-7ad3-7cd1-a524-a39a93d90450):

> I probably already said that - I don't like the idea of 3-stage process. This proposal was accepted and now reflects how SocialWG operates in practice:
>
>https://swicg.github.io/charters/stage-process
>
>According to the document, FEPs are at the same level as blog posts. Even a successful and widely implemented FEP needs a "champion" who will sign [CLA](https://socialhub.activitypub.rocks/t/fep-5219-groups-and-permissions/8855/8) and will present the FEP to the community group, which will bikeshed it to death. Even if it passes the initial filter of the community group, the document still needs to be reviewed by the working group, which is invite-only.
>
>This is a ridiculous amount of red tape.
>
>Compare this to the promise of the FEP process: you publish a FEP -> you convince other developers -> it becomes a de-facto standard.

And I [followed up](https://web.archive.org/web/20260830170733/https://social.coop/@smallcircles/117185510811483948) with:

>Having looked in more detail at the [#SWICG](https://social.coop/tags/SWICG) document. That's not the 3-stage process I advocated for at the time, and I agree with you: It will not work. Such top-down formal ceremony just does not fit a grassroots ecosystem.
>
>Generally speaking I don't think [#W3C](https://social.coop/tags/W3C) as a standardization body is up to the task of being a good host for the ActivityPub specification. It is a) not organized for our environment, and b) only takes care of the technical aspects. Our social media landscape has a huge social component, and this is wholly underrepresented in our app-centric fediverse, where people are merely "users" and [#FOSS](https://social.coop/tags/FOSS) culture is the only driving force.
>
>Today I think SWICG should focus on the 'Protocol implementer' stakeholder: How can they offer compliant [#ActivityPub](https://social.coop/tags/ActivityPub) protocol implementations.
>
>Then [#FEP](https://social.coop/tags/FEP) can focus on the 'Solution developer', relieved from the plumbing: What interoperable designs do they use to serve the needs of participants on the social network.

**Comment by evanp** [comment]
> While this issue is still open, a more elaborate [staged process](https://swicg.github.io/charters/stage-process) is being proposed.

I think it was actually accepted when the WG charter was approved. So, this is the agreed-upon way that the CG and the WG work together.

> I want to mention feedback wrt this document, by [@silverpill](https://github.com/silverpill) and also by me, on the fediverse..

Why do you want to mention it? To what end?

> > According to the document, FEPs are at the same level as blog posts.

Yes, from the point of the view of the W3C, FEPs, blog posts, ISO standards and United Nations Security Council resolutions are all somewhat equivalent: documents produced outside the CG and the WG that may or may not be used as [normative references](https://www.w3.org/guide/process/tilt/normative-references.html) in W3C specifications.

I think there's an [extant resolution](https://www.w3.org/2021/01/09-social-minutes.html#t02) by the CG to support the FEP process, which is probably the main way that FEPs differ from other external documents.

> > Compare this to the promise of the FEP process: you publish a FEP -> you convince other developers -> it becomes a de-facto standard.

That's absolutely fine. Clearing issues like patent licensing and other IP grants -- the things that the CG's CLA is for -- is necessary for some documents and not others. It's perfectly reasonable for a lot of developers to use FEPs out of the box.

Some FEPs can be input to CG [Specifications](https://www.w3.org/community/about/process/#deliverables). Some might go directly into CG Specifications, with some light formatting changes as required by the W3C.

Some CG Specifications will be input to WG [Recommendations](https://www.w3.org/policies/process/#rec-track).

The staging process is about how the WG and CG decide together which documents we publish.

> > Generally speaking I don't think [#W3C](https://social.coop/tags/W3C) as a standardization body is up to the task of being a good host for the ActivityPub specification.

I strongly disagree. The W3C has been an excellent steward for ActivityPub.

> > Today I think SWICG should focus on the 'Protocol implementer' stakeholder: How can they offer compliant [#ActivityPub](https://social.coop/tags/ActivityPub) protocol implementations.
> > Then [#FEP](https://social.coop/tags/FEP) can focus on the 'Solution developer', relieved from the plumbing: What interoperable designs do they use to serve the needs of participants on the social network.

I don't understand the difference you're positing here. Could you give an example or two?

Ultimately, it's up to the CG what the CG will do. There's a pretty good [scope of work](https://swicg.github.io/charters/SocialWebCG-Charter-2025-02.html#scope-of-work) in the charter. If you think that should change, you should bring it up with the CG, either on the mailing list or in a call.

**Comment by aschrijver** [comment]
> Why do you want to mention it? To what end?

Because for years and years I have dedicated myself to try to help improve the developer ecosystem. It is wicked hard, this grassroots environment we are faced with. The fact that currently docs document the process, does not mean the process is in place and functioning well. I am not in the opportunity to spend much time on this now, but I want to leave this feedback, in the place where there is an open issue for it. Simple as that.

The SWICG is supposedly the representative of the entire ActivityPub ecosystem. In practice that isn't the case, and the ecosystem is fully grassroots and fragmented. SWICG operates as a loose group, and there are other loose groups, like FEP. And individual devs who take their info from anywhere. It is all one big grassroots chaos. There effectively is no bottom-up process, and nowhere is an authoritative place in this ecosystem. In order for the staged process to become meaningful, the SWICG needs to get more buy-in, and there needs to be streamlined collaboration with those other places.

I think SWICG should be the most authoritative body, but that only works if it is accepted as such, not if it "writes" its place in the hierarchy with top-down ceremony.

There are people who say that W3C ActivityPub Recommendation is a good protocol specification. But it simply isn't. Instead it is a _good start_ of such a specification. At best ActivityPub as-is is a good _framework_ for making protocol specifications. Handing that out to devs for many years has created a cacophony of AP flavors, and delivered a wholly app-centric fediverse dominated by the most popular AP flavor creators.

> I don't understand the difference you're positing here. Could you give an example or two?

If I am a protocol implementer, I should be able to go to the SWICG and AS/AP specs, work through them, and deliver a protocol implementation I can hand to any solution developer. Implementing my protocol implementation should be straightforward, where I find all the necessary plumbing details to make it work properly. And I should have good test suites to verify that.

As a solution developer I should be able to take such AP protocol implementation and immediately continue to worry only about the solution I want to build. Microblogging? Let's go. Video processing? Gotcha. Instead any federated app developer is forever mired in technical intricacies that constitutes protocol plumbing, simply because there isn't much consensus on how things ought to be done. The lack of good specs and guidance results in a work method of looking what others do and copy/paste mechanisms from all over the place, then commence to do whack-a-mole programming and maintenance to eternity to fix random breakages caused by a remote app.

This morning I added [following toot](https://web.archive.org/web/20260831083629/https://social.coop/@smallcircles/117189061613262222) to the thread, that relates to this as well:

>Yes, this is the eternal problem in AP fediverse space. And it causes most devs to leave the ecosystem again after dabbling in it for a while, getting mired in endless low-level plumbing which they might be shielded from if there were robust specs to work with.
>
>I think if anything this should be the major major concern that the SWICG should dedicate to and tackle once and for all. This is the task that is not picked up since 2018. And gain buy-in on that in the larger ecosystem, so there's more unity among developers.
>
>This is what I meant with focus on the protocol implementer to give them what they need: Great specs to follow, to create compliant AP protocol implementations. And then there are countless FEP's to go all different counts of directions and explore particular use cases and designs.

**Comment by aschrijver** [comment]
> I strongly disagree. The W3C has been an excellent steward for ActivityPub.

My observation is that the open standard has stagnated since 2018 during which time a whole different reality has emerged in the installed base of the fediverse wich evolved itself based on the incomplete standard that was available. During that time a tremendous amount of protocol decay and tech debt has been accepted, which makes it ultra hard today to fix that mess.

The W3C as standardization body works best for industry technology standards, where technology partners from various companies collaborate on web standards. Their formal standardization process works in those situations, but it is not a process that fits the grassroots social dynamics in the fediverse ecosystem. Furthermore it is exclusively tech-focused, and has none of the social components that perhaps the effort to shape alternative social media solutions to Big Tech requires, where the broader fediverse community also has more say in the evolution of the field, other than just app developers.

But I guess the question boils down to "What do we (collectively) want the fediverse to be?", and that discussion isn't happening. It is all just tech to build upon as one wishes. That is bound to end up in a bad place, imho.

**Comment by evanp** [comment]
OK. I think you're proposing a change to the SocialCG scope, which I think is possible. The scope is laid out in the charter. There's a section on amending the charter here:

https://swicg.github.io/charters/SocialWebCG-Charter-2025-02.html#charter-change

I think the next step is starting that proposal. I think the chairs @jernst and @dmitrizagidulin would be able to help with that.

I disagree with you, but I think you should get to say your piece.

**Comment by aschrijver** [comment]
Thanks. As mentioned I am not sure how much time I can spend atm, as I am in desperate need of income and busy with that.

**Comment by evanp** [comment]
> Thanks. As mentioned I am not sure how much time I can spend atm, as I am in desperate need of income and busy with that.

Ah, that's too bad. Thanks for sharing your thoughts!

**Comment by bumblefudge** [comment]
> Compare this to the promise of the FEP process: you publish a FEP -> you convince other developers -> it becomes a de-facto standard.

Sure, sounds good-- there are probably more time-effective ways of convincing other developers!  A de facto standard (or in this case, a de facto _profile_, to be pedantic) doesn't care what W3C thinks-- only browser vendors and global platforms do.  

Let me reframe this a little, as the editor of this process doc.

There is no obligation for a FEP author to champion their features/changes/ideas in the WG unless they want it to be "upstreamed"/mainlined into some future version of the spec.  And there is no obligation for a FEP author to champion their FEP in the CG unless they want to convince other implementers through dialogue about implementation details/side-effects/ramifications, which can (but does not need to) take the form of task forces or reports or email threads.  It's all entirely opt-in, and the least bureaucratic version of CG participation is just sending an email to the list that says, "here is a FEP/blogpost/codeberg link describing what I do in my implementation.  Other implementers who want to discuss can email me directly about it."  That's actually my _favorite_ kind of participation, because it still _protects_ the implementer (and innovation more generally) from the CG and WG without requesting consensus of either.

I will say, conversely, that I consider a published FEP implemented by one still-maintained, used codebase as significant documentary evidence that a given feature or specified behavior of the existing spec HAS been used, which is very valuable in W3C process. If that codebase is broken by any spec changes (even ones classed as "errata" on version 1.0.1 of the spec), I will lobby against that breakage on behalf of that implementation, no CLA or participation needed on the part of the implementer.  The tricky part here is avoiding breakage of implementations we don't know about!
